zeppelin-zengine/src/main/java/org/apache/zeppelin/users/UsersRepo.java

```diff
@@ -0,0 +1,170 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.zeppelin.users;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.nio.file.Files;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import static java.nio.file.attribute.PosixFilePermission.OWNER_READ;
+import static java.nio.file.attribute.PosixFilePermission.OWNER_WRITE;
+
+/**
+ * Class for storing user information (persistence layer).
+ */
+public class UsersRepo {
+  private static final Logger LOG = LoggerFactory.getLogger(UsersRepo.class);
+
+  private Gson gson;
+  private File usersFile;
+
+  private Map<String, UserInfo> usersInfo = new HashMap<>();
+
+  private boolean persistMode;
+
+  public UsersRepo() {
+    persistMode = false;
+  }
+
+  public UsersRepo(String usersInfoPath) {
+    persistMode = true;
+    if (usersInfoPath != null) {
+      this.usersFile = new File(usersInfoPath);
+    }
+
+    GsonBuilder builder = new GsonBuilder();
+    builder.setPrettyPrinting();
+    gson = builder.create();
+    loadFromFile();
+  }
+
+  public UserInfo getUserInfo(String user) {
+    UserInfo userInfo = usersInfo.get(user);
+    if (userInfo == null) {
+      userInfo = new UserInfo();
+    }
+    return userInfo;
+  }
+
+  public void putRecentNote(String user, String noteId) throws IOException {
+    synchronized (usersInfo) {
+      if (!usersInfo.containsKey(user)) {
+        usersInfo.put(user, new UserInfo());
+      }
+      usersInfo.get(user)
+          .addRecentNote(noteId);
+      saveToFileIfPersistMode();
+    }
+  }
+
+  public void removeNoteFromRecent(String user, String noteId) throws IOException {
+    synchronized (usersInfo) {
+      usersInfo.get(user).removeNoteFromRecent(noteId);
+      saveToFileIfPersistMode();
+    }
+  }
+
+  /**
+   * Remove note from recent of all users
+   * @param noteId - id of removing note
+   * @throws IOException if problems with save to file
+   */
+  public void removeNoteFromRecent(String noteId) throws IOException {
+    synchronized (usersInfo) {
+      for (Map.Entry<String, UserInfo> e : usersInfo.entrySet()) {
+        e.getValue().removeNoteFromRecent(noteId);
+      }
+      saveToFileIfPersistMode();
+    }
+  }
+
+  public void clearRecent(String user) throws IOException {
+    synchronized (usersInfo) {
+      usersInfo.get(user).clearRecent();
+      saveToFileIfPersistMode();
+    }
+  }
+
+  private void loadFromFile() {
```

Comment:
As we store more and more configuration, the number of configuration file increases. This introduces some problems

  - A high cost on development.
    Each developer need to implement load/store function when a feature need to store some informations. Also this new configuration file creation needs went through discussion/review process.
  - High I/O overhead.
    Like @zjffdu mentioned, depends on how frequently information should be stored, in some case it may introduce high I/O overhead.
  - Not really following standard convention for *nix application
    Many *nix application stores configuration under /conf or /config dir.
    And it's common system administrator does not give edit (or sometimes both edit/read) permission of those configuration files to the user.
   However, Zeppelin mixes 'human' edit configuration (like conf/zeppelin-site.xml) and machine stored configuration (like interpreter.json, helium.json) in the same directory.
  That makes system administrator difficult to manage permission of two different configuration files.
   https://github.com/apache/zeppelin/pull/2539 is one of an example of the problem.

@zjffdu @tinkoff-dwh How about we create a separate issue for the problem related to configuration files and address separately, while this PR focus on implementing recent list feature?